src/modules/bulletproofs/main_impl.h

```diff
@@ -0,0 +1,628 @@
+    secp256k1_ge value_genp;
+    secp256k1_ge **commitp;
+    size_t i;
+    int ret;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(scratch != NULL);
+    ARG_CHECK(proof != NULL);
+    ARG_CHECK(circ != NULL);
+    ARG_CHECK(gens != NULL);
+    ARG_CHECK(gens->n >= 2 * circ[0]->n_gates);
+    ARG_CHECK(value_gen != NULL);
+    ARG_CHECK((extra_commit_len == NULL) == (extra_commit == NULL));
+    if (extra_commit != NULL) {
+        for (i = 0; i < n_proofs; i++) {
+            ARG_CHECK(extra_commit[i] != NULL || extra_commit_len[i] == 0);
+        }
+    }
+
+    for (i = 1; i < n_proofs; i++) {
+        ARG_CHECK(circ[i]->n_gates == circ[0]->n_gates);
+    }
+
+    if (n_commits == NULL) {
+        commitp = NULL;
+    } else {
+        size_t total_n_commits = 0;
+        for (i = 0; i < n_proofs; i++) {
+            total_n_commits += n_commits[i];
+        }
+
+        if (!secp256k1_scratch_allocate_frame(scratch, sizeof(*commitp) + total_n_commits * sizeof(**commitp), 1 + n_proofs)) {
+            return 0;
+        }
+        commitp = (secp256k1_ge **)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*commitp));
+        for (i = 0; i < n_proofs; i++) {
+            size_t j;
+            commitp[i] = (secp256k1_ge *)secp256k1_scratch_alloc(scratch, n_commits[i] * sizeof(*commitp[i]));
+            for (j = 0; j < n_commits[i]; j++) {
+                secp256k1_pedersen_commitment_load(&commitp[i][j], &commit[i][j]);
+            }
+        }
+    }
+
+    secp256k1_generator_load(&value_genp, value_gen);
+    ret = secp256k1_bulletproof_relation66_verify_impl(
+        &ctx->ecmult_ctx,
+        scratch,
+        proof, n_proofs, plen,
+        (const secp256k1_ge* const*)commitp, n_commits,
+        &value_genp,
+        circ,
+        gens,
+        extra_commit, extra_commit_len
+    );
+    if (n_commits != NULL) {
+        secp256k1_scratch_deallocate_frame(scratch);
+    }
+    return ret;
+}
+
+secp256k1_bulletproof_circuit *secp256k1_bulletproof_circuit_decode(const secp256k1_context *ctx, const char *fname) {
+    FILE *fh;
+    unsigned char buf[0x3f];
+    size_t n;
+    size_t row_width;
+    secp256k1_bulletproof_circuit *ret;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(fname != NULL);
+
+    fh = fopen(fname, "rb");
+    if (fh == NULL) {
+        return NULL;
+    }
+
+    /* Read header and allocate memory */
+    if (secp256k1_bulletproof_circuit_allocate_memory(ctx, fh, &ret) != 1) {
+        fclose(fh);
+        return NULL;
+    }
+    row_width = secp256k1_bulletproof_encoding_width(ret->n_gates);
```

Comment:
shouldn't row_width be dependent on the `n_constraints` instead of `n_gates`?  Every entry in a row encodes the index of the constraint (also encoded with `row_width`-many bytes) the wire is added to and the factor the wire is multiplied with in that constraint. Therefore there are at most `n_constraint` many entries in a row.